### DISCUSSION TITLE
Experiment with RGBX in libjpeg

### DISCUSSION THREAD

**Discussion by jcupitt** [libvips/libvips#2028] [Ideas]
If we enable RGBX from libjpeg we could potentially get easier vectorization in reduceh. 

See https://github.com/libvips/libvips/pull/816#issuecomment-346899199

**Comment by kleisauke** [comment]
I had a very quick (and dirty) attempt: https://github.com/kleisauke/libvips/tree/add-rgbx.

It's able to load and save with the `JCS_EXT_RGBX` color space, but it marks a JPEG image loaded via libjpeg-turbo as a 4-band `VIPS_INTERPRETATION_RGBX` image. I was unable to tag it as a 3-band image without making a lot of exclusions for this interpretation.

Perhaps it makes more sense to create a new `VIPS_CODING_RGBX`? This allows us to tag it as a normal 3-band `VIPS_INTERPRETATION_sRGB` image.

**Comment by jcupitt** [comment]
`pngload` writes an INTERPRETATION_sRGB image with 4 bands, would that work?

**Comment by jcupitt** [comment]
You're right, a CODING_RGBX would be useful for images which are strictly locked to 32-bit RGBX pixels, but just tagging as RGB 4 bands might be enough for a simple proof of concept benchmark.

**Comment by kleisauke** [comment]
I tried the micro-benchmarks from https://github.com/kleisauke/vips-microbench with the above mentioned branch, but saw no difference (I'll push the results soon). Perhaps this should be tested on a large image with shrink-on-load disabled.

The vector path within `reduceh.cpp` should probably also be enabled to properly benchmark this.
https://github.com/libvips/libvips/blob/72c103f95a3407ce35bdfdb78d189ceae8a26098/libvips/resample/reduceh.cpp#L292-L293

**Comment by kleisauke** [comment]
I tried to create a vector path for `reduceh`, but it was harder than I thought. Do you still have the code for the experiment mentioned in [`3c98ba7`](https://github.com/libvips/libvips/commit/3c98ba78a0469c74ba378889aa4089aed08d2ec2#diff-d81f64fd1fa063fc9364e0d3572d6e70)?

By the way,
While comparing between `reduceh` and `reducev`, I think I might have found a little copy-paste error:
```diff
diff --git a/libvips/resample/reduceh.cpp b/libvips/resample/reduceh.cpp
index 1111111..2222222 100644
--- a/libvips/resample/reduceh.cpp
+++ b/libvips/resample/reduceh.cpp
@@ -172,12 +172,14 @@ reduceh_unsigned_int_tab( VipsReduceh *reduceh,
 
 	for( int z = 0; z < bands; z++ ) {
 		int sum;
-	       
-		sum = reduce_sum<T, int>( in + z, bands, cx, n );
+
+		sum = reduce_sum<T, int>( in, bands, cx, n );
 		sum = unsigned_fixed_round( sum ); 
 		sum = VIPS_CLIP( 0, sum, max_value ); 
 
 		out[z] = sum;
+
+		in += 1;
 	}
 }
 
```

I immediately checked whether issue 703 had been solved with this, but unfortunately that was not the case.

**Comment by kleisauke** [comment]
I managed to create a vector path within `reduceh` (see https://github.com/kleisauke/libvips/commit/8235dfab0527dcefee3913729a536249515bd89e). Unfortunately I don't see a big difference in the vectorized `reduceh` and my current implementation of the RGBX mode. The benchmark results can be viewed here:
https://github.com/kleisauke/vips-microbench/blob/master/RESULTS.md

Note that these benchmarks can be seen as proof-of-concept, I'll try a few more experiments to see if performance can be improved.

**Comment by kleisauke** [comment]
Small update: the vector path wasn't working correctly (it fell back to the non-vectorised method). Fixed with: https://github.com/kleisauke/libvips/commit/44f1602a359474e6476c978e3a01ec0097f3e414.

The benchmark results have been adjusted to reflect this change.
